src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1976,6 +1977,58 @@ multi_relation_restriction_hook(PlannerInfo *root, RelOptInfo *relOptInfo,
 }
 
 
+/*
+ * Normally, we should not need this. However, the combination of
+ * Postgres commit 3c569049b7b502bb4952483d19ce622ff0af5fd6 and
+ * Citus function AdjustPartitioningForDistributedPlanning()
+ * forces us to do this. The commit expects partitioned indexes
+ * to belong to relations with "inh" flag set properly. Whereas, the
+ * function overrides "inh" flag. To avoid a crash,
+ * we go over the list of indexinfos and remove all partitioned indexes.
+ * Partitioned indexes were ignored pre PG16 anyway, we are essentially
+ * not breaking any logic.
+ *
+ * AdjustPartitioningForDistributedPlanning() is a hack that we use
+ * to prevent Postgres' standard_planner() to expand all the partitions
+ * for the distributed planning when a distributed partitioned table
+ * is queried. It is required for both correctness and performance
+ * reasons. Although we can eliminate the use of the function for
+ * the correctness (e.g., make sure that rest of the planner can handle
+ * partitions), it's performance implication is hard to avoid. Certain
+ * planning logic of Citus (such as router or query pushdown) relies
+ * heavily on the relationRestrictionList. If
+ * AdjustPartitioningForDistributedPlanning() is removed, all the
+ * partitions show up in the, causing high planning times for
+ * such queries.
+ */
+void
+multi_partitioned_index_hook(PlannerInfo *root, Oid relationObjectId, bool inhparent,
+							 RelOptInfo *rel)
+{
+	Index varno = rel->relid;
+	RangeTblEntry *rangeTableEntry = planner_rt_fetch(varno, root);
+
+	if (rangeTableEntry->rtekind == RTE_RELATION &&
+		PartitionedTable(rangeTableEntry->relid) &&
+		rangeTableEntry->inh == false)
+	{
+		LOCKMODE lmode = root->simple_rte_array[varno]->rellockmode;
+		List *indexinfos = NIL;
+		IndexOptInfo *indexOptInfo = NULL;
+		foreach_ptr(indexOptInfo, rel->indexlist)
+		{
+			Relation indexRelation = index_open(indexOptInfo->indexoid, lmode);
+			if (indexRelation->rd_rel->relkind != RELKIND_PARTITIONED_INDEX)
+			{
+				indexinfos = lappend(indexinfos, indexOptInfo);
+			}
+			index_close(indexRelation, lmode);
+		}
+		rel->indexlist = indexinfos;
```

Comment:
`indexinfos` is allocated in this memory context, but `rel->indexlist` is probably allocated in another memory context.  We cannot assume that both have the same context.

We should switch to the context that `rel->indexlist` is in, and then do `lappend` in that context.

Please see `GetMemoryChunkContext` and its usage.

Reply:
or, maybe we should properly delete from the list?

Reply:
> or, maybe we should properly delete from the list?

I tried with list_delete_cell, is that what you meant?
I haven't pushed it yet since I ran into some problems with running it.

Reply:
I mean, if we add `IsCitusTable()` I probably wouldn't mind doing `rel->indexlist = NIL;`, 

In the end, we don't care about the indexes on the coordinating node?

Reply:
> yeah, list_delete_cell is tricky, but I meant that

I found what is not tricky: `foreach_delete_current` :)

> I mean, if we add IsCitusTable() I probably wouldn't mind doing rel->indexlist = NIL;

I still have to find a way to process non-Citus tables as well. AdjustPartitioning function is also adjusting regular Postgres tables. So, I'd rather not do that, but keep a general method for all tables. This would also preserve the behavior we had until PG16.